artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPConnectionCallback.java

```diff
@@ -113,7 +116,20 @@ public ServerSASL getServerSASL(final String mechanism) {
                result = gssapiServerSASL;
                break;
 
+            case ExternalServerSASL.NAME:
+               // validate ssl cert present
+               Principal principal = CertificateUtil.getPeerPrincipalFromConnection(protonConnectionDelegate);
+               if (principal != null) {
+                  ExternalServerSASL externalServerSASL = new ExternalServerSASL();
+                  externalServerSASL.setPrincipal(principal);
+                  result = externalServerSASL;
+               } else {
+                  logger.debug("SASL EXTERNAL mechanism requires a TLS peer principal");
```

Comment:
This feels like it should be an exception rather than just a log message. Things shouldn't get here if the connection cant actually do EXTERNAL, since the server shouldn't offer it in that case (as it can tell before offering that it cant work), and should fail before here if the client selected it when it wasn't actually offered.

Reply:
that is fair. there is currently not much logic around what mechanisms are supported, if a list is configured it is returned to the client.
The result of returning null here fails at

https://github.com/apache/activemq-artemis/pull/1961/files/c2869ca6598b7d17a56aee451daebfb7cb01fa0b#diff-f80f5c57a928d9c39f6dd7f7ea8028dfR316